compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -1207,7 +1207,7 @@ impl<'tcx> ParamEnv<'tcx> {
             Reveal::UserFacing => ParamEnvAnd { param_env: self, value },
 
             Reveal::All => {
-                if value.is_global() {
+                if value.is_global() && !value.has_ty_fresh() {
```

Comment:
hmm, i would assume that `has_ty_fresh` should instead be a subset of `is_global`. i.e. i would add `HAS_TY_FRESH` and `HAS_CT_FRESH` to `HAS_FREE_LOCAL_NAMES` :thinking:

Reply:
After some further thought, I think you're correct. I originally decided against doing that because I was worried about affecting unrelated calls of `is_global()`. However, it should always be correct to return `false` from `is_global`, since `is_global() == true` is used for caching and bypassing checks.